README.md
# js-password-generator
a fully featured password genorator that can be used as a web app or as part of another application, such as generating passwords for account resets or as part of a password manager.

This application is still in development, at the bottom are a list of features desired for the app and which ones are complete.

There is a live version of this app at http://www.mhzsys.net/applications/js-password-generator/

This generator is based on and inspired by the password generator at, 
https://passwordsgenerator.net/

The app has the ability to chose what type of characters to use with several customization options.  The use of Upper Case, Lower Case, And numbers is use by default.

To call the function simply use generatePassword({_length: 8}); with the first paramater being a number stating how many characters long you want the password to be.  Other options are available below and can be added by sperating them in the function with a comma ,

- _length: #number  -Determines length of password
- _lower: true/false        -If you want to use lower case letters
- _lowerMin: #number        -If you want a mimimum ammount of lower case letters used
- _upper: true/false        -If you want to use upper case letters
- _upperMin: #number        -If you want a mimimum ammount of upper case letters used
- _numb: true/false         -If you want to use numbers
- _numbMin: #number         -If you want a minimum ammount of numbers to be used
- _special: true/false      -If you want to use special characters [!, @, #, $, %, ^, &, *, (, )]
- _specialMin: #number      -If you want a minimum ammount of special characters to be used
- _optional: true/false     -If you want to use optional characters [-, _, =, +, [, ], {, }, ;, :, '', ", <, >, ,, ., ?, /, |, `, ~]
- _optionalMin: #number     -If you want a minimum ammount of optional characters to be used
- _firstLetter: true/false  -If you want the first character to be a letter
- _repeat: true/false       -If you do not want two of the same character next to each other
- _noSimilar: true/false    -If you do not want to use similar looking characters.
- _noSequential: true/false -If you do not want characters that are normally next to each other to be used as such, 123, abc, etc.


# Features
- [X]Choose password length.
- [X]Choose lowercase.
- [X]Choose uppercase.
- [X]Choose numbers.
- [X]Choose special characters.
- [X]Choose other special characters.
- [X]Prevent repeats.
- [X]Ensure first character is a letter.
- [X]Exclude similar characters. (helpful if the password is not being coppied and pasted)
- [X]Basic presets if options are not selected.
- [X]Don't use sequential characters.
- [ ]Choose specific characters to exclude.
- [X]Spefify a minimum ammount for each character type.
- [ ]Password strength checker.

# Bugs and fixes
- [X]Prohibbit combination of no uppercase, no lowercase, begins with a letter. (causes crash)
- [X]Prevent minimum characters from exceeding password length.
- [X]Find a better way to specify options while calling the function.

# Similar characters
These characters are often mistaken for each other if being typed or especially if being written down.
o O 0 i j l I S $ 5 B 8

changelog.md
# Change Log

# 2018-08-11
- Added minimum type of characters functionality.
- Fixed bug to prevent user from selecting more minimum characters then password length.
- Fixed bug to prevent user from de selecting letters and requiring first character to be a letter.

v1/controls.js
'use strict';

//function to generate password to web page
document.getElementById("generate-password").addEventListener("click", function (event) {
  event.preventDefault();

  //choose character options and customization
  let passwordLength = document.getElementById("password-length").value;
  let passwordLowerCase = document.getElementById("lower-case").checked;
  let passwordLowerCaseMin = document.getElementById("lower-case-min").value;
  let passwordUpperCase = document.getElementById("upper-case").checked;
  let passwordUpperCaseMin = document.getElementById("upper-case-min").value;
  let passwordNumbers = document.getElementById("numbers").checked;
  let passwordNumbersMin = document.getElementById("number-min").value;
  let passwordSpecialCharacters = document.getElementById("special-characters").checked;
  let passwordSpecialMin = document.getElementById("special-case-min").value;
  let passwordOptionalCharacters = document.getElementById("optional-characters").checked;
  let passwordOptionalMin = document.getElementById("optional-case-min").value;
  let passwordBeginsWithLetter = document.getElementById("begin-with-letter").checked;
  let passwordRepeatingCharacters = document.getElementById("repeating-characters").checked;
  let passwordNoSimilarCharacters = document.getElementById("similar-characters").checked;
  let passwordNoSequentialCharacters = document.getElementById("sequential-characters").checked;

  let result = generatePassword(
    passwordLength,
    passwordLowerCase,
    passwordLowerCaseMin,
    passwordUpperCase,
    passwordUpperCaseMin,
    passwordNumbers,
    passwordNumbersMin,
    passwordSpecialCharacters,
    passwordSpecialMin,
    passwordOptionalCharacters,
    passwordOptionalMin,
    passwordBeginsWithLetter,
    passwordRepeatingCharacters,
    passwordNoSimilarCharacters,
    passwordNoSequentialCharacters
  );
  document.getElementById("generated-password").value = result;
});

password-generator.js
'use strict';

//constant values to use with password.
const lowerCase = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']; //26 characters
const upperCase = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']; //26 characters
const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']; //10 characters
const specialCharacters = ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')']; //10 Characters
const optionalCharacters = ['-', '_', '=', '+', '[', ']', '{', '}', ';', ':', '\'', '\"', '<', '>', ',', '.', '?', '/', '|', '`', '~']; //21 Characters
const similarCharacters = ['o', 'O', '0', 'i', 'j', 'l', 'I', 'S', '$', '5', 'B', '8'];

let meetsReqirements = true;
//function to generate random password
function generatePassword({
  _length,
  _lower,
  _lowerMin,
  _upper,
  _upperMin,
  _numb,
  _numbMin,
  _special,
  _specialMin,
  _optional,
  _optionalMin,
  _firstLetter,
  _repeat,
  _noSimilar,
  _noSequential
} = {}) {
  do {
    meetsReqirements === true
    // debugger;
    let passwordLength = _length;
    let lower = _lower;
    let lowerMin = _lowerMin;
    let upper = _upper;
    let upperMin = _upperMin;
    let numb = _numb;
    let special = _special;
    let specialMin = _specialMin;
    let optional = _optional;
    let optionalMin = _optionalMin;
    let firstLetter = _firstLetter;
    let repeat = _repeat;
    let noSimilar = _noSimilar;
    let noSequential = _noSequential;
    let numbMin = _numbMin;


    let upperCheck = 0;
    let lowerCheck = 0;
    let numbCheck = 0;
    let specialCheck = 0;
    let optionalCheck = 0;

    let generatedPassword = [];
    let parameters = [];

    let minCharacters = parseInt(lowerMin) + parseInt(upperMin) + parseInt(numbMin) + parseInt(specialMin) + parseInt(optionalMin)
    
    // Bug fix for min characters and password length.
    if (minCharacters > parseInt(passwordLength)){
      console.log('Minimum ammount of characters can not exceed requestd password length.')
      alert('Minimum ammount of characters can not exceed requestd password length.')
      break;
    }

    // Bug fix for no letters while requiring pass to begin with a letter.
    if(lower == false && upper == false && firstLetter == true){
      console.log('Must use Letters if requiring first character to be a letter.')
      alert('Must use Letters if requiring first character to be a letter.')
      break;
    }

    //set lower case by default
    if (lower == false) {
      //do nothing
    } else if (lower == null || true) {
      parameters.push(lowerCase);
    };

    //set upper case by default
    if (upper == false) {
      //do nothing
    } else if (upper == null || true) {
      parameters.push(upperCase);
    };

    //set numbers case by default
    if (numb == false) {
      //do nothing
    } else if (numb == null || true) {
      parameters.push(numbers);
    };

    //enables special characters
    if (special == true) {
      parameters.push(specialCharacters);
    }

    //enables optional characters
    if (optional == true) {
      parameters.push(optionalCharacters);
    }


    // console.log('parameters: ', parameters)
    for (let i = 0; i < passwordLength; i++) {
      let charSet = Math.floor(Math.random() * parameters.length);
      let charSel = Math.floor(Math.random() * parameters[charSet].length);

      //Checking password for speficied modifications

      if (repeat == true) {
        if (i == 0) {
          console.log('no duplicate characters selected.');
        }
        if (parameters[charSet][charSel] === generatedPassword[i - 1]) {
          console.log('duplicate character found and dropped');
          generatedPassword.pop();
          i--;
        }
      }
      
      generatedPassword.push(parameters[charSet][charSel]);

      // if enabled will make sure first character is a letter.
      if (firstLetter == true) {
        // need something to prevent selecting false on upper and lower case to cause crash.

        let first = false;
        if (i == 0) {
          console.log('first letter option selected.');
          for (let j = 0; j < lowerCase.length; j++) {
            if (generatedPassword[0] == lowerCase[j] || generatedPassword[0] == upperCase[j]) {
              first = true;
            }
          }
          if (first == false) {
            console.log('First character not a letter');
            generatedPassword.pop();
            i--;
          }
        }
      }

      // if enabled will not use similar looking characters.
      if (noSimilar == true) {
        if (i == 0) {
          console.log('no similar characters option selected.');
        }
        for (let k = 0; k < similarCharacters.length; k++) {
          if (generatedPassword[i] == similarCharacters[k]) {
            console.log('Found invalid character');
            generatedPassword.pop();
            i--;
          }
        }
      }

      // if enabled will not use sequential characters.
      if (noSequential == true) {
        if (i == 0) {
          console.log('no sequential characters option selected.');
        }
        if (i > 0) {
          if (generatedPassword[i - 1] === parameters[charSet][charSel - 1] || generatedPassword[i - 1] === parameters[charSet][charSel + 1]) {
            generatedPassword.pop();
            i--;
          }
        }
      }
    };
    // console.log(generatedPassword);
    // Sets minimum characters to be used.
    for (let i = 0; i < generatedPassword.length; i++) {
      for (let j = 0; j < lowerCase.length; j++) { if (generatedPassword[i] === lowerCase[j]) { lowerCheck++; } };
      for (let k = 0; k < upperCase.length; k++) { if (generatedPassword[i] === upperCase[k]) { upperCheck++; } };
      for (let l = 0; l < numbers.length; l++) { if (generatedPassword[i] === numbers[l]) { numbCheck++; } };
      for (let m = 0; m < specialCharacters.length; m++) { if (generatedPassword[i] === specialCharacters[m]) { specialCheck++; } };
      for (let n = 0; n < optionalCharacters.length; n++) { if (generatedPassword[i] === optionalCharacters[n]) { optionalCheck++; } };
    }
    // console.log('upperCheck: ', upperCheck, 'lowerCheck: ', lowerCheck, 'numbCheck: ', numbCheck, 'specialCheck: ', specialCheck, 'optionalCheck: ', optionalCheck)
    if (lower == true && lowerMin > lowerCheck) {
      meetsReqirements = false;
      console.log('lower requested: ', lowerMin, 'result: ', lowerCheck)
    }else

    if (upper == true && upperMin > upperCheck) {
      meetsReqirements = false;
      console.log('upper requested: ', upperMin, 'result: ', upperCheck)
    }else

    if (numb == true && numbMin > numbCheck) {
      meetsReqirements = false;
      console.log('number requested: ', numbMin, 'result: ', numbCheck)
    }else

    if (special == true && specialMin > specialCheck) {
      meetsReqirements = false;
      console.log('special requested: ', specialMin, 'result: ', specialCheck)
    }else

    if (optional == true && optionalMin > optionalCheck) {
      meetsReqirements = false;
      console.log('optional requested: ', optionalMin, 'result: ', optionalCheck)
    }else{
      meetsReqirements = true
    }



    if (meetsReqirements == true) {
      return generatedPassword.join('');
    } else {
      console.log('does not meet: ', generatedPassword.join(''))
      generatedPassword = {};
      // reRunPasswordGeneration(passwordLength, lower, lowerMin, upper, upperMin, numb, special, specialMin, optional, optionalMin, firstLetter, repeat, noSimilar, noSequential, numbMin)
    };

  }
  while (meetsReqirements == false);
}

function reRunPasswordGeneration(passwordLength, lower, lowerMin, upper, upperMin, numb, special, specialMin, optional, optionalMin, firstLetter, repeat, noSimilar, noSequential, numbMin) {
  // console.log('PW re run: ', passwordLength, lower, lowerMin, upper, upperMin, numb, special, specialMin, optional, optionalMin, firstLetter, repeat, noSimilar, noSequential, numbMin)
  generatePassword(passwordLength, lower, lowerMin, upper, upperMin, numb, special, specialMin, optional, optionalMin, firstLetter, repeat, noSimilar, noSequential, numbMin)
}

app.js
'use strict';

console.log("app.js loaded");

//constant values to use with password.
const lowerCase = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']; //26 characters
const upperCase = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']; //26 characters
const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']; //10 characters
const specialCharacters = ['!', '@', '#', '$', '%', '^', '&', '*', '(', ')']; //10 Characters
const optionalCharacters = ['-', '_', '=', '+', '[', ']', '{', '}', ';', ':', '\'', '\"', '<', '>', ',', '.', '?', '/', '|', '`', '~']; //21 Characters

//function to generate random password
function generatePassword(passwordLength, lower, upper, numb, special, optional, firstLetter, repeat) { //passwordLength determines how many characters the password will be.
  let generatedPassword = [];
  let parameters = [];

  //set lower case by default
  if (lower == false) {
    //do nothing
  } else if (lower == null || true) {
    parameters.push(lowerCase);
  };

  //set upper case by default
  if (upper == false) {
    //do nothing
  } else if (upper == null || true) {
    parameters.push(upperCase);
  };

  //set numbers case by default
  if (numb == false) {
    //do nothing
  } else if (numb == null || true) {
    parameters.push(numbers);
  };

  //enables special characters
  if (special == true) {
    parameters.push(specialCharacters);
  }

  //enables optional characters
  if (optional == true) {
    parameters.push(optionalCharacters);
  }

  for (let i = 0; i < passwordLength; i++) {
    let charSet = Math.floor(Math.random() * parameters.length);
    let charSel = Math.floor(Math.random() * parameters[charSet].length);

    //Checking password for speficied modifications

    if (repeat == true) {
      if (parameters[charSet][charSel] === generatedPassword[i - 1]) {
        console.log('duplicate character found and dropped');
        generatedPassword.pop();
        i--;
      }
    }

    generatedPassword.push(parameters[charSet][charSel]);

    // if enabled will make sure first character is a letter.
    if (firstLetter == true) {
      // need something to prevent selecting false on upper and lower case to cause crash.

      let first = false;
      if (i == 0) {
        console.log('first letter option selected.');
        for (let j = 0; j < lowerCase.length; j++) {
          if (generatedPassword[0] == lowerCase[j] || generatedPassword[0] == upperCase[j]) {
            first = true;
          }
        }
        if(first == false) {
          console.log('First character not a letter');
          generatedPassword.pop();
          i--;
        }
      }
    }
  };
  console.log('gen pass length: ', generatedPassword.length);
  return generatedPassword.join('');
};

//function to generate password to web page
document.getElementById("generate-password").addEventListener("click", function (event) {
  event.preventDefault();

  //choose character options and customization
  let passwordLowerCase = document.getElementById("lower-case").checked;
  let passwordUpperCase = document.getElementById("upper-case").checked;
  let passwordNumbers = document.getElementById("numbers").checked;
  let passwordSpecialCharacters = document.getElementById("special-characters").checked;
  let passwordOptionalCharacters = document.getElementById("optional-characters").checked;
  let passwordBeginsWithLetter = document.getElementById("begin-with-letter").checked;
  let passwordRepeatingCharacters = document.getElementById("repeating-characters").checked;

  let passwordLength = document.getElementById("password-length").value;
  let result = generatePassword(
    passwordLength,
    passwordLowerCase,
    passwordUpperCase,
    passwordNumbers,
    passwordSpecialCharacters,
    passwordOptionalCharacters,
    passwordBeginsWithLetter,
    passwordRepeatingCharacters
  );
  document.getElementById("generated-password").value = result;
});